### ISSUE TITLE
How to check if option exists?

### ISSUE THREAD

**Issue by ZedZipDev** [commandlineparser/commandline#369]
Is it possible to check if this option passed by user or it is only the default value?

**Comment by ericnewton76** [comment]
I don't believe so.  An implementation redesign could hold that as a list of options and values that were set by the library to determine.  

A better way to determine via your own implementation is to keep a record of property setter calls on your Options type: (psuedo code, so excuse syntax errors)

```c#
public class MyOptions 
{
    private string _MyOption1 = "some_default_value";
    [Option("MyOption1")]
    public string MyOption1 
    { 
        get => _MyOption1;
        set {
            _SetValues[nameof(MyOption1)] = value;
            _MyOption1 = value;
        }

    private Dictionary<string,string> _SetValues = new Dictionary<string,string>();
    public IDictionary<string,string> GetSetValues() => _SetValues;
}
```

Now you'll know if the library set the value or if it was never set, thus staying as the default:
```cmd
> myprogram --MyOption1 some_default_value
```

**Comment by Brice-xCIT** [comment]
Unless I missed a configuration gotcha, @ericnewton76's solution does not seem to work anymore with 2.6.0. Properties are always set to their default value even if the Option was not present in the arguments.

**Comment by moh-hassan** [comment]
@Brice-xCIT 
>Properties are always set to their default value even if the Option was not present in the arguments.

Properties are set to `default(T)` [See Default values table](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/default-values-table)

For `Nullable<T>` it's set to null.

Value is overwritten by `Default Property of OptionAttribute` if it's set.
```cs
        [Option(Default=123)]
        public int Size { get; set; } //Size will be 123 if the Option was not present in the arguments
```

**Comment by Brice-xCIT** [comment]
Hi @moh-hassan, I understand your point, however the issue is not to give an option a default value, but to identify if an option was set by the user or not. In other words, was the option populated with a default, or was it populated with the user input? There could be different ways of doing that but currently the library offers none.

**Comment by moh-hassan** [comment]
@Brice-xCIT 
You can use Unparser Extension method `FormatCommandLine` with the new feature in v2.7 [SkipDefault ](https://github.com/commandlineparser/commandline/wiki/Unparsing#skipping-options-with-default-values) to know if the option  is populated with a default or set by user.
The following Extension method `IsSetByUser` is used to check the option status.

```cs	
  static class Ex
{
  // return true if option is set by user otherwise it return false;
  public static bool IsSetByUser<T>(this T options, string option)
  {
  //unparse options
  	string args = CommandLine.Parser.Default.FormatCommandLine(options, config => config.SkipDefault = true);
	 Console.WriteLine($"Unparser:  {args}");   
	 var status =args.Contains(option);
	 if (status)
	 {
		 Console.WriteLine($"option: {option} is Set by user");
		 return true;
	 }
	 Console.WriteLine("option: {option} is Set by Parser");	
	return false;
  }
}  
```

**Comment by Brice-xCIT** [comment]
Hi @moh-hassan, thank you for pointing out to this new feature, it looks promising. However if the user did pass a value that was also the default value, this method will not detect it. I'm thinking now maybe the best solution is to just use `Nullable<>` on properties that need to be tracked for user customization.
